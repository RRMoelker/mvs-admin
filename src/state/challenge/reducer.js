import { createAction } from 'redux-actions';

// import { reduxSwarmLogId } from '../actions.js';

import { v4 } from 'uuid';

import { mergeOverlapping } from './merger.js';
import {
    calculateActive,
    calculateRecent,
    calculateRemaining,
    groupChallenges,
    getRemaining
} from './compute.js';

const CHALLENGE_ADD = 'CHALLENGE_ADD';
// export const addChallenge = createAction(CHALLENGE_ADD);
export const addChallenge = payload => ({
    type: CHALLENGE_ADD,
    payload,
    // reduxSwarmLogId
});

const CHALLENGE_REMOVE = 'CHALLENGE_REMOVE';
// export const removeChallenge = createAction(CHALLENGE_REMOVE);
export const removeChallenge = payload => ({
    type: CHALLENGE_REMOVE,
    payload,
    // reduxSwarmLogId
});

const initialState = {
    list: [],
}

export const selectRemaining = (list, now) => {
    const grouped = groupChallenges(list);

    const remaining = {};
    for( const [ key, groupList ] of Object.entries(grouped)) {
        const merged = mergeOverlapping(groupList);
        const active = calculateActive(merged, now);
        if ( active.length ) {
            remaining[key] = calculateRemaining(active, now);
        }
    }
    return remaining;
}

export const selectRecent = (list, now, threshold) => {
    const recentList = calculateRecent(list, now, threshold);
    const grouped = groupChallenges(recentList);
    const result = {};
    for( const [ key, value ] of Object.entries(grouped) ) {
        result[key] = value.reduce(
            (acc, val) => acc + val.duration,
            0
        );
    };
    return result;
}

export default (state = initialState, action) => {
    switch (action.type) {
        case CHALLENGE_ADD:
            const { name, duration } = action.payload;
            const { time } = action.meta;

            const newChallenge = {
                uuid: v4(),
                name,
                duration,
                time,
            };
            const list = [ ...state.list, newChallenge];
            list.sort((a,b) => {
                return a.time > b.time;
            });

            return {
                ...state,
                list,
            }
        case CHALLENGE_REMOVE:
            const filteredList = state.list.filter( item => item.uuid !== action.payload.uuid);
            return {
                ...state,
                list: filteredList
            };
        default:
            return state;
    }
};
