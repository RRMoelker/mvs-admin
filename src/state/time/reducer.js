import { createAction } from 'redux-actions';

export const TIME_SET = 'TIME_SET';
export const setTime = createAction(TIME_SET);
const TIME_RESET = 'TIME_RESET';
export const resetTime = createAction(TIME_RESET);
const TIME_PAUSE = 'TIME_PAUSE';
export const pauseTime = createAction(TIME_PAUSE);

const initialState = {
    time: 0,
    running: true
}

export default (state = initialState, action) => {
    switch (action.type) {
        case TIME_RESET:
            return {
                ...state,
                time: 0,
            }
        case TIME_SET:
            return {
                ...state,
                time: action.payload.time,
            }
        case TIME_PAUSE:
            return {
                ...state,
                running: !state.running
            }
        default:
            return state;
    }
}
