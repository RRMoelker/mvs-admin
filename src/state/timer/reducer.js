import { createAction } from 'redux-actions';

export const TIMER_SET = 'TIMER_SET';
export const setTimer = createAction(TIMER_SET);
const TIMER_RESET = 'TIMER_RESET';
export const resetTimer = createAction(TIMER_RESET);
const TIMER_PAUSE = 'TIMER_PAUSE';
export const pauseTimer = createAction(TIMER_PAUSE);

const initialState = {
    time: 0,
    running: true
}

export default (state = initialState, action) => {
    switch (action.type) {
        case TIMER_RESET:
            return {
                ...state,
                time: 0,
            }
        case TIMER_SET:
            return {
                ...state,
                time: action.payload.time,
            }
        case TIMER_PAUSE:
            return {
                ...state,
                running: !state.running
            }
        default:
            return state;
    }
}
