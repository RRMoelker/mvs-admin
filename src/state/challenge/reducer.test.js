import reducer, {
    addChallenge,
    removeChallenge,
    selectRemaining
} from './reducer.js';

describe('the challenge reducer', () => {
    it('should start with an empty list', () =>{
        const result = reducer(undefined, { type: 'NONE' });

        expect(result).toMatchSnapshot();
    });

    it('should push new challenges to the list', () =>{
        const action = addChallenge({
            name: 'minimap',
            duration: 1000
        });
        action.meta = { time: 0 };
        const actionB = addChallenge({
            name: 'glove',
            duration: 2000
        });
        actionB.meta = { time: 1 };

        const stepState = reducer(undefined, action);
        const result = reducer(stepState, actionB);

        expect(result).toMatchSnapshot();
    });
});

describe('the challenge reducer', () => {
    it('should remove elements from the list', () =>{
        const action = addChallenge({
            name: 'minimap',
            duration: 1000
        });
        action.meta = { time: 0 };
        const stepState = reducer(undefined, action);

        const result = reducer(stepState, removeChallenge({ uuid: stepState.list[0].uuid}));

        expect(result.list.length).toBe(0);
        expect(result).toMatchSnapshot();
    });
});

describe('select remaining challenges', () => {
    it('should calculate a single challenge', () =>{
        const action = addChallenge({
            name: 'glove',
            duration: 10
        });
        action.meta = { time: 1 };
        const state = reducer(undefined, action);

        const list = state.list;
        const result = selectRemaining(list, 3);

        expect(result).toEqual({
            glove: {
                remaining: 8,
                until: 11
            }
        });
    });

    it('should handle two challenges', () => {
        const action = addChallenge({
            name: 'glove',
            duration: 10
        });
        action.meta = { time: 5 };
        const actionB = addChallenge({
            name: 'glove',
            duration: 10
        });
        actionB.meta = { time: 20 };

        const stateA = reducer(undefined, action);
        const stateB = reducer(stateA, actionB);

        const list = stateB.list;
        const result = selectRemaining(list, 25);

        expect(result).toEqual({
            glove: {
                remaining: 5,
                until: 30
            }
        });
    });

    it('should handle overlapping challenges', () => {
        const action = addChallenge({
            name: 'glove',
            duration: 10
        });
        action.meta = { time: 5 };
        const actionB = addChallenge({
            name: 'glove',
            duration: 10
        });
        actionB.meta = { time: 10 };
        const stateA = reducer(undefined, action);
        const stateB = reducer(stateA, actionB);

        const list = stateB.list;
        const result = selectRemaining(list, 12);

        expect(result).toEqual({
            glove: {
                remaining: 13,
                until: 25
            }
        });
    });

    it('should be empty if all challenges have passed', () => {
        const action = addChallenge({
            name: 'glove',
            duration: 10
        });
        action.meta = { time: 0 };
        const actionB = addChallenge({
            name: 'glove',
            duration: 10
        });
        actionB.meta = { time: 20 };
        const actionC = addChallenge({
            name: 'glove',
            duration: 10
        });
        actionC.meta = { time: 40 };
        const stateA = reducer(undefined, action);
        const stateB = reducer(stateA, actionB);
        const stateC = reducer(stateB, actionC);

        const list = stateC.list;
        const result = selectRemaining(list, 51);

        expect(result).toEqual({});
    });

    it('should support mixed challenges', () => {
        const action = addChallenge({
            name: 'glove',
            duration: 10
        });
        action.meta = { time: 0 };
        const actionB = addChallenge({
            name: 'map',
            duration: 10
        });
        actionB.meta = { time: 15 };
        const actionC = addChallenge({
            name: 'health',
            duration: 10
        });
        actionC.meta = { time: 18 };
        const stateA = reducer(undefined, action);
        const stateB = reducer(stateA, actionB);
        const stateC = reducer(stateB, actionC);

        const list = stateC.list;
        const result = selectRemaining(list, 19);

        expect(result).toEqual({
            map: {
                remaining: 6,
                until: 25
            },
            health: {
                remaining: 9,
                until: 28
            }
        });
    });
});
