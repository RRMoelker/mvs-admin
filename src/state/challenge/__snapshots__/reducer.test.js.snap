// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the challenge reducer the challenge reducer should push new challenges to the list 1`] = `
Object {
  "list": Array [
    Object {
      "duration": 1000,
      "name": "minimap",
      "time": 0,
      "uuid": 0,
    },
    Object {
      "duration": 2000,
      "name": "glove",
      "time": 1,
      "uuid": 1,
    },
  ],
}
`;

exports[`the challenge reducer the challenge reducer should remove elements from the list 1`] = `
Object {
  "list": Array [],
}
`;

exports[`the challenge reducer the challenge reducer should start with an empty list 1`] = `
Object {
  "list": Array [],
}
`;
